onnxruntime/core/providers/cuda/tensor/nonzero_op.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "nonzero_op.h"
+#include "nonzero_impl.h"
+#include "core/providers/cpu/tensor/utils.h"
+#include "core/providers/common.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+// kernel builder functions
+#define NONZERO_TYPED_KERNEL_WITH_TYPE_NAME(type, type_name)                       \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                                   \
+      NonZero,                                                                     \
+      kOnnxDomain,                                                                 \
+      9,                                                                           \
+      type_name,                                                                   \
+      kCudaExecutionProvider,                                                      \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<type>()), \
+      NonZero<type>)
+
+#define NONZERO_TYPED_KERNEL(type) \
+  NONZERO_TYPED_KERNEL_WITH_TYPE_NAME(type, type)
+
+// start with a subset of types, enable more as needed...
+NONZERO_TYPED_KERNEL(bool)
+NONZERO_TYPED_KERNEL(uint8_t)
+//NONZERO_TYPED_KERNEL(uint16_t)
+//NONZERO_TYPED_KERNEL(uint32_t)
+//NONZERO_TYPED_KERNEL(uint64_t)
+//NONZERO_TYPED_KERNEL(int8_t)
+//NONZERO_TYPED_KERNEL(int16_t)
+NONZERO_TYPED_KERNEL(int32_t)
+NONZERO_TYPED_KERNEL(int64_t)
+NONZERO_TYPED_KERNEL(MLFloat16)
+//NONZERO_TYPED_KERNEL(BFloat16)
+NONZERO_TYPED_KERNEL(float)
+//NONZERO_TYPED_KERNEL(double)
+//NONZERO_TYPED_KERNEL_WITH_TYPE_NAME(std::string, string)
+
+#undef NONZERO_TYPED_KERNEL
+#undef NONZERO_TYPED_KERNEL_WITH_TYPE_NAME
+
+template <typename T>
+Status NonZero<T>::ComputeInternal(OpKernelContext* context) const {
+  static const std::vector<int64_t> kScalarDims = {1};
+  const auto x = context->Input<Tensor>(0);
+
+  int nonzero_elements = 0;
+  const auto& x_shape = x->Shape();
+  const int x_rank = x_shape.IsScalar() ? 1 : static_cast<int>(x_shape.NumDimensions());
```

Comment:
>static_cast [](start = 46, length = 11)

Not for this PR, but we may consider /permissive- in MSVC to catch it locally before hitting CI

[https://devblogs.microsoft.com/cppblog/permissive-switch/](https://devblogs.microsoft.com/cppblog/permissive-switch/)

Reply:
Cool! Will read it.